operator/pkg/ciliumidentity/cache_test.go

```diff
@@ -0,0 +1,294 @@
+		return
+	}
+
+	labelsfilter.ParseLabelPrefixCfg(nil, "")
+	testLabelsFilterInitialized = true
+}
+
+func TestCIDState(t *testing.T) {
+	testInitLabelsFilter()
+	state := NewCIDState()
+
+	k := GetCIDKeyFromK8sLabels(k8sLables_A)
+	state.Upsert("1", k)
+
+	expectedState := &CIDState{
+		idToLabels: map[string]*key.GlobalIdentity{"1": k},
+		labelsToID: map[string]*SecIDs{
+			k.GetKey(): {
+				selectedID: "1",
+				ids:        map[string]bool{"1": true},
+			},
+		},
+	}
+
+	assert.NoError(t, validateCIDState(state, expectedState), "cid 1 added")
+
+	k = GetCIDKeyFromK8sLabels(k8sLables_B)
+	state.Upsert("2", k)
+
+	expectedState.idToLabels["2"] = k
+	expectedState.labelsToID[k.GetKey()] = &SecIDs{
+		selectedID: "2",
+		ids:        map[string]bool{"2": true},
+	}
+
+	assert.NoError(t, validateCIDState(state, expectedState), "cid 2 added")
+
+	k = GetCIDKeyFromK8sLabels(k8sLables_B_duplicate)
+	state.Upsert("3", k)
+
+	expectedState.idToLabels["3"] = k
+	expectedState.labelsToID[k.GetKey()] = &SecIDs{
+		selectedID: "2",
+		ids:        map[string]bool{"2": true, "3": true},
+	}
+
+	assert.NoError(t, validateCIDState(state, expectedState), "cid 3 added - duplicate")
+
+	cidKey, exists := state.LookupByID("0")
+	assert.Equal(t, false, exists, "cid 0 LookupByID - not found")
+
+	cidKey, exists = state.LookupByID("1")
+	assert.Equal(t, true, exists, "cid 1 LookupByID - found")
+	assert.Equal(t, GetCIDKeyFromK8sLabels(k8sLables_A), cidKey, "cid 1 LookupByID - correct key")
+
+	_, exists = state.LookupByKey(GetCIDKeyFromK8sLabels(k8sLables_C))
+	assert.Equal(t, false, exists, "labels C LookupByKey - not found")
+
+	cidName, exists := state.LookupByKey(GetCIDKeyFromK8sLabels(k8sLables_A))
+	assert.Equal(t, true, exists, "labels C LookupByKey - not found")
+	assert.Equal(t, "1", cidName, "labels C LookupByKey - correct CID")
+
+	state.Remove("2")
+
+	delete(expectedState.idToLabels, "2")
+	expectedState.labelsToID[GetCIDKeyFromK8sLabels(k8sLables_B).GetKey()] = &SecIDs{
+		selectedID: "3",
+		ids:        map[string]bool{"3": true},
+	}
+
+	assert.NoError(t, validateCIDState(state, expectedState), "cid 2 removed")
+
+	cidKey, exists = state.LookupByID("2")
+	assert.Equal(t, false, exists, "cid 2 LookupByID - not found")
+
+	state.Remove("3")
+	delete(expectedState.idToLabels, "3")
+	delete(expectedState.labelsToID, GetCIDKeyFromK8sLabels(k8sLables_B_duplicate).GetKey())
+
+	assert.NoError(t, validateCIDState(state, expectedState), "cid 3 removed")
+}
+
+func TestCIDStateThreadSafety(t *testing.T) {
+	// This test ensures that no changes to the CID state break its thread safety.
+	// Multiple go routines in parallel continuously keep using CIDState.
+	testInitLabelsFilter()
+	state := NewCIDState()
+
+	k := GetCIDKeyFromK8sLabels(k8sLables_A)
+	k2 := GetCIDKeyFromK8sLabels(k8sLables_B)
+
+	wg := sync.WaitGroup{}
+	queryStateFunc := func() {
+		startTime := time.Now()
+
+		for time.Since(startTime) < 500*time.Millisecond {
```

Comment:
I'd prefer to run a fixed number of operations instead of "as much as possible in a given time interval", because the timing characteristic of the CI may vary greatly, making the test flaky.

Reply:
Suggestion accepted.